### ISSUE TITLE
ngc watch is not incremental, runs full build

### ISSUE THREAD

**Issue by alippai** [angular/angular#21361]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
To compile my project it takes 11 seconds for tsc (typescript@2.6), however ngc takes 25 seconds.
The watch mode is similar 3 seconds versus 10+ for a single file (component) change.
Is there a way to debug the problem or is this the desired behavior. 

## Expected behavior
The durations should be close

## Environment

<pre><code>
Angular version: 5.2.0-rc0
TypeScript version (both global and project): 2.6

For Tooling issues:
- Node version: 8.9.3 and 9.3.0
- Platform: Both Windows and the Linux subsystem (Ubuntu)

tsconfig:
{
  "compilerOptions": {
    "target": "ES2017",
    "module": "ES2015",
    "moduleResolution": "node",
    "sourceMap": false,
    "importHelpers": true,
    "emitDecoratorMetadata": false,
    "experimentalDecorators": true,
    "lib": [
      "es2015",
      "dom",
      "es2017.object"
    ],
    "noImplicitAny": false,
    "isolatedModules": true,
    "suppressImplicitAnyIndexErrors": true,
    "allowSyntheticDefaultImports": true,
    "allowJs": true,
    "skipLibCheck": true,
    "outDir": "../../compiled",
    "typeRoots": [
      "../../node_modules/@types/"
    ]
  },

  "files": [
    "./app.module.ts"
  ],

  "include": [
    "../../../../types/models/*.d.ts"
  ],

  "exclude": [
    "../../generated/**",
    "../../compiled/**",
    "../../node_modules/**"
  ]
}
</code></pre>

**Comment by alippai** [comment]
I can't get the diagnostics for the ngc, the tsc is here:
```
$ ./node_modules/.bin/tsc -p app/ts/tsconfig.watch.json --diagnostics
Files:          1186
Lines:        100855
Nodes:        411168
Identifiers:  125264
Symbols:      119669
Types:         21711
Memory used: 174866K
I/O read:      0.20s
I/O write:     0.37s
Parse time:    2.48s
Bind time:     0.79s
Check time:    2.94s
Emit time:     1.01s
Total time:    7.22
```
The same file & options:
```
$ ./node_modules/.bin/ngc -p app/ts/tsconfig.watch.json --diagnostics
(empty line)
```

**Comment by alippai** [comment]
https://github.com/angular/angular/blob/83d207d0a795bc1945c4661cb5ab73d2e671c1ce/packages/compiler-cli/src/perform_watch.ts#L198

The oldProgram option below is _always_ undefined.

**Comment by alippai** [comment]
After the fix, it's still fishy, 200 files changed by changing only 1 angular component:
```
C:\Users\Lippai Adam\Repositories\webclient\app\admin>node ./node_modules/@angular/compiler-cli/src/main.js -p app/ts/tsconfig.watch.json -w
: Time for diagnostics: 10057ms.
: Emitted in 10920ms
- 1270 user ts files
- 249 generated ts files
- 860 generated json files

: Total time: 21s

: Compilation complete. Watching for file changes.

: File change detected. Starting incremental compilation.

: Time for diagnostics: 7021ms.
: Emitted in 7045ms
- 1343 user ts files
- 99 generated ts files
- 99 generated json files

: Total time: 14s

: Compilation complete. Watching for file changes.
```

**Comment by alippai** [comment]
3rd and 4th runs are better, but still slow:
```
$ node ./node_modules/\@angular/compiler-cli/src/main.js -p app/tsconfig.watch.json -w
: Time for diagnostics: 10827ms.
: Emitted in 10415ms
- 1270 user ts files
- 249 generated ts files
- 860 generated json files

: Total time: 21s

: Compilation complete. Watching for file changes.

: File change detected. Starting incremental compilation.

: Time for diagnostics: 7426ms.
: Emitted in 7579ms
- 1343 user ts files
- 99 generated ts files
- 99 generated json files

: Total time: 15s

: Compilation complete. Watching for file changes.

: File change detected. Starting incremental compilation.

: Time for diagnostics: 7040ms.
: Emitted in 3325ms
- 1 user ts files
- 0 generated ts files
- 0 generated json files

: Total time: 10s

: Compilation complete. Watching for file changes.

: File change detected. Starting incremental compilation.

: Time for diagnostics: 6198ms.
: Emitted in 4309ms
- 1 user ts files
- 0 generated ts files
- 0 generated json files

: Total time: 11s

: Compilation complete. Watching for file changes.
```

**Comment by alippai** [comment]
The last change results in smaller generated code, which is unexpected for me. @IgorMinar or @tbosch who is familiar with this piece of code?

**Comment by alippai** [comment]
@chuckjaz how should I proceed? Should I rebase continuously or is there anything I can help the merging process?

**Comment by alippai** [comment]
What's happening? I've asked for a simple comment after fixing two bugs and there is no answer. If you have other personal or project preferences just add it and feel free to close the issue without explanation.

**Comment by IgorMinar** [comment]
@alippai that was just our bot helping us to categorize issues based on their current state.

the compiler is currently not fully incremental because of some architectural constraints that we are removing in the new renderer we call ivy. we should however investigate if your change really improves the current situation. we'll need @chuckjaz to take a look at your PR as he's the one who needs to approve the change in this piece of code.

sorry about the delay. I'll see if we can have chuck take a quick look next week.

**Comment by alippai** [comment]
Thank you for your detailed answer. I'm happy to help, just didn't know
that should I proceed or there is something in your mind. Feel free to to
ask for updates. Ivy really looks good, keep up the good work!

P.S I wanted to say, that this project could use a PM intern, who helps us
whether we should contribute or not. It's not about you or Chuck, the work
you do is amazing.

**Comment by IgorMinar** [comment]
We are trying to automate most of the team and contributor workflows, so
I'm hoping that PRs like yours will get better attention faster without us
reviewing the same PR many times in hopes that it has been updated and
tests pass. Thanks for being patient.

**Comment by alippai** [comment]
@IgorMinar any updates? One of the two PRs is a single word typo, helps to make the watch a magnitude faster.

**Comment by alippai** [comment]
This is not something what should happen in open source. Maintaining code quality and having strategic goals in development is one thing, the intentional total lack of communication is an other. Not just not being able to fix a clear typo affecting the only compiler feature mentioned in Angular 5's release blog post (in 6 months!), but ignoring a contributors very first PR. This is not welcoming or inclusive. Better I stick to MS+FB+other open source projects, but not Google's next time.

I really appreciate your hard work, the amount of expertise and work you share is wonderful and unique. You really help everybody. It's just not the open source project I expected it to be.

**Comment by sod** [comment]
angular-cli or @ngtools/webpack doesn't rely on this code, so I guess `ngc --watch` is kind of in a blind spot.

They have written off this compiler already as they write a new compiler (ivy) that can run file2file instead of managing the hole project. At least they have to if they want to leverage the power of bazel.

Lets hope ivy works just like typescript. Can compile file by file and run diagnostics in a separate thread. So compile time comes done to milliseconds instead of seconds.

**Comment by alxhub** [comment]
Closing as fixed. @alippai's fix was merged in c6e5b971.
